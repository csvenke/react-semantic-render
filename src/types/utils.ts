import * as PropTypes from 'prop-types';
import * as React from 'react';

export type Map =
  | ((item?: any, index?: number, array?: any[]) => React.ReactNode)
  | null;

export type Render = (() => React.ReactNode) | null;

export type Children = React.ReactNode | (() => React.ReactNode) | null;

export interface IRenderProps {
  /** Shorthand for primary content. */
  render?: Render;

  /** Primary content. */
  children?: Children;
}

export const renderPropsPropTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  render: PropTypes.func,
};
