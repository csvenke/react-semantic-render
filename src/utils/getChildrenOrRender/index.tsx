import * as React from 'react';

import { isEmptyChildren } from '../isEmptyChildren';

export const getChildrenOrRender = (children?: any, render?: () => React.ReactNode) => {
  if (typeof children === 'function') {
    return children();
  }

  if (children && !isEmptyChildren(children)) {
    return React.Children.only(children);
  }

  if (render) {
    return render();
  }

  return null;
};
