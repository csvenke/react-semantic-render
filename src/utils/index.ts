export * from './isEmptyChildren';
export * from './isPromise/isPromise';
