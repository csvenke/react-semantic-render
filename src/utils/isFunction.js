function isFunction(input) {
  return typeof input === "function";
}

export default isFunction;
