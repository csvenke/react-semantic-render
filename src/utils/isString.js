function isString(input) {
  return typeof input === "string";
}

export default isString;
