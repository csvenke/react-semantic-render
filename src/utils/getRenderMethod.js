import isFunction from './isFunction'

function getRenderMethod (children, render) {
  if (children) {
    return children
  }

  if (render && isFunction(render)) {
    return render
  }

  return null
}

export default getRenderMethod
