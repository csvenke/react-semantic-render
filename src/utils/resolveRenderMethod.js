import React from "react";

import isEmptyChildren from "./isEmptyChildren";
import isFunction from "./isFunction";
import isString from "./isString";

function resolveRenderMethod(method) {
  if (isFunction(method)) {
    return method.call();
  }

  if (!isEmptyChildren(method) && React.isValidElement(method)) {
    return React.Children.only(method);
  }

  if (isString(method)) {
    return method;
  }

  return null;
}

export default resolveRenderMethod;
