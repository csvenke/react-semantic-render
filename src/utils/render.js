import getRenderMethod from "./getRenderMethod";
import resolveRenderMethod from "./resolveRenderMethod";

function render(props) {
  const method = getRenderMethod(props.children, props.render);
  return resolveRenderMethod(method);
}

export default render;
