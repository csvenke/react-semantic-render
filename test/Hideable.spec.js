import React from 'react'
import renderer from 'react-test-renderer'

import { createTests } from './utils'
import { Hideable as Hideable1 } from '../dist/index.cjs'
import { Hideable as Hideable2 } from '../dist/index.esm'
import Hideable3 from '../dist/Hideable'
import Hideable4 from '../src/Hideable'

const runTests = createTests(Hideable1, Hideable2, Hideable3, Hideable4)

runTests(Hideable => {
  const Button = () => <button>Click me</button>
  const HideableButton = Hideable(Button)

  test('should show component', () => {
    const component = renderer.create(
      <HideableButton hide={false} label='Click me!' />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('should return null', () => {
    const component = renderer.create(
      <HideableButton hide label='Click me!' />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
