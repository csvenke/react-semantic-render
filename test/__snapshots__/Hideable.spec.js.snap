// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return null 1`] = `null`;

exports[`should return null 2`] = `null`;

exports[`should return null 3`] = `null`;

exports[`should return null 4`] = `null`;

exports[`should show component 1`] = `
<button>
  Click me
</button>
`;

exports[`should show component 2`] = `
<button>
  Click me
</button>
`;

exports[`should show component 3`] = `
<button>
  Click me
</button>
`;

exports[`should show component 4`] = `
<button>
  Click me
</button>
`;
